### PR TITLE
OCR

### LINKED ISSUE #55: Execute OCR on the scanned documents
We should connect Tesseract OCR to implement full text search on the scanned books and other documents.

### PR DESCRIPTION
**Pull request by yuliiabuchko**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -71,6 +71,8 @@ dependencies {
     compile("org.glassfish.jaxb:jaxb-runtime:2.3.2")
     compile("commons-io:commons-io:2.6")
     compile("org.apache.tika:tika-parsers:1.24")
+    compile("org.apache.pdfbox:pdfbox:2.0.1")
```
Comment (apardyl) [inline build.gradle]: outdated?

File: build.gradle [hunk 2]
```diff
@@ -71,6 +71,8 @@ dependencies {
...
+    compile("net.sourceforge.tess4j:tess4j:3.4.0")
```
Comment (apardyl) [inline build.gradle]: outdated?

File: gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -1,5 +1,6 @@
+#Mon May 04 21:03:43 CEST 2020
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.2-all.zip
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.2-bin.zip
-zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
```
Comment (apardyl) [inline gradle/wrapper/gradle-wrapper.properties]: revert?

File: src/main/kotlin/pl/edu/uj/ii/ksi/mordor/configuration/TesseractConfiguration.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import net.sourceforge.tess4j.Tesseract
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+
+@Configuration
+class TesseractConfiguration {
+    @Bean
+    fun tesseract(): Tesseract {
+        val tesseract = Tesseract()
+        tesseract.setDatapath("./src/main/resources/")
```
Comment (apardyl) [inline src/main/kotlin/pl/edu/uj/ii/ksi/mordor/configuration/TesseractConfiguration.kt]: check if it works when mordor is bundled in a release jar file

File: src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/PDFTextExtractor.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        for (page in doc.pages) {
+            val resources = page.resources
+            for (xObjectName in resources.xObjectNames) {
+                val xObject = resources.getXObject(xObjectName)
+                if (xObject is PDImageXObject) {
+                    bufferedImages.add(xObject.image)
+                }
+            }
+        }
+        doc.close()
```
Comment (apardyl) [inline src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/PDFTextExtractor.kt]: avoid manually closing streams, use the `use` keyword

File: src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/ImageTextExtractor.kt [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        return ImageHelper.convertImageToGrayscale(image)
+    }
+
+    fun extractTextFromBufferedImage(image: BufferedImage?): String? {
+        try {
+            return tesseract.doOCR(correctTwisted(image))
+                    .replace("\\n{2,}", "\n")
+                    .trim { c -> c <= ' ' }
+        } catch (e: TesseractException) {
+            LoggerFactory.getLogger(this.javaClass).error("Tesseract is unable do process OCR on image", e)
```
Comment (apardyl) [inline src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/ImageTextExtractor.kt]: make the logger a companion object (so it is created only on when the class is loaded)

**Review by apardyl — CHANGES_REQUESTED** [review]

File: src/main/kotlin/pl/edu/uj/ii/ksi/mordor/configuration/TesseractConfiguration.kt (hunk 4 above)
Comment (yuliiabuchko) [reply src/main/kotlin/pl/edu/uj/ii/ksi/mordor/configuration/TesseractConfiguration.kt]: I've checked, it works

File: src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/text/extractor/ImageTextExtractor.kt [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import org.bytedeco.tesseract.TessBaseAPI
+
+class ImageTextExtractor(private val tessBaseAPI: TessBaseAPI) : FileTextExtractor {
+
+    override fun extract(file: File, maxLength: Int): String? {
+        tessBaseAPI.SetImage(pixRead(file.absolutePath))
+
+        val res = tessBaseAPI.GetUTF8Text().string.trimIndent()
+
+        return if (maxLength >= 0 && maxLength < res.length) {
```
Comment (apardyl) [inline src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/text/extractor/ImageTextExtractor.kt]: use String::take

File: src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/text/extractor/FileContentValidator.kt [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        if (content.isNullOrEmpty()) {
+            return true
+        }
+        return whiteSpaceFilter(content)
+    }
+
+    private fun whiteSpaceFilter(content: String): Boolean {
+        val letters = content.filter { c -> c.isLetterOrDigit() }.length
+        if (letters.toFloat().div(content.length) < minAlphanumericsPercent) {
+            logger.warn("Number of alphanumeric chars is less than 50%. OCR result will be turned to null")
```
Comment (apardyl) [inline src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/text/extractor/FileContentValidator.kt]: you can probably easily make it stronger - the average word length in English language is 4.7 characters.

**Review by apardyl — CHANGES_REQUESTED** [review]
looks almost ok, please manually check if it works on pdf files from mordor.ksi.ii.uj.edu.pl

File: src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/text/extractor/AutoDetectTextExtractor.kt [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+    private fun isScanned(content: String?): Boolean {
+        return content?.trim()?.isEmpty() ?: true
+    }
+
+    private fun cleanSmallWords(content: String?): String? {
+        if (content == null) {
+            return null
+        }
+        val modified = content.replace('\n', ' ').replace('\t', ' ')
```
Comment (apardyl) [inline src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/text/extractor/AutoDetectTextExtractor.kt]:
```suggestion
        val modified = content.replace("\\s".toRegex(), " ")
```

File: src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/text/extractor/AutoDetectTextExtractor.kt [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            if (type == "application/pdf") {
+                logger.info("Extracted text from " + file.absolutePath + " using Bytedeco for PDF")
+                return PDFTextExtractor(tessBaseAPI).extract(file, maxLength)
+            }
+            if (type.startsWith("image")) {
+                logger.info("Extracted text from " + file.absolutePath + " using Bytedeco")
+                return ImageTextExtractor(tessBaseAPI).extract(file, maxLength)
+            }
+        } catch (e: IOException) {
+            logger.error("File can not be read", e)
```
Comment (apardyl) [inline src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/text/extractor/AutoDetectTextExtractor.kt]:
```suggestion
            logger.warn("File can not be read: " + file.absolutePath, e)
```

**Review by apardyl — CHANGES_REQUESTED** [review]
looks ok, just some nits

**Review by apardyl — APPROVED** [review]

**Pull request merged**